lib/rmg/CMapGenerator.cpp

```diff
@@ -389,23 +396,41 @@ int CMapGenerator::getNextMonlithIndex()
 
 int CMapGenerator::getPrisonsRemaning() const
 {
-	return prisonsRemaining;
+	return allowedPrisons;
 }
 
-void CMapGenerator::decreasePrisonsRemaining()
+std::shared_ptr<CZonePlacer> CMapGenerator::getZonePlacer() const
 {
-	prisonsRemaining = std::max (0, prisonsRemaining - 1);
+	return placer;
 }
 
-const std::vector<ArtifactID> & CMapGenerator::getQuestArtsRemaning() const
+const std::vector<ArtifactID> & CMapGenerator::getAllPossibleQuestArtifacts() const
 {
 	return questArtifacts;
 }
 
+const std::vector<HeroTypeID> CMapGenerator::getAllPossibleHeroes() const
+{
+	//Skip heroes that were banned, including the ones placed in prisons
+	std::vector<HeroTypeID> ret;
+	for (int j = 0; j < map->map().allowedHeroes.size(); j++)
+	{
+		if (map->map().allowedHeroes[j])
+			ret.push_back(HeroTypeID(j));
+	}
+	return ret;
+}
+
 void CMapGenerator::banQuestArt(const ArtifactID & id)
 {
+	//TODO: Protect with mutex
```

Comment:
If two threads can simultaneously generate two quests this might lead to RMG producing different maps from the same seed/settings.

Although not sure if we need generation of identical maps in RMG via settings same seed, so feel free to ignore.

Reply:
That's why there is this task:

> Give every Zone private random number generator

Each zone will generate exactly the same regardless their order. There is only a risk that two zones may try to place an artifact in third zone at the same time, and then the order will be undefined. But artifacts on map are randomized at game start anyway, so it doesn't change much.